File: src/transport/connection.rs
use std::net::SocketAddr;
use std::sync::atomic::{AtomicU32, Ordering};
use std::time::{Duration, Instant};

use crate::protocol::types::{ConnectionState, StreamConfig, StreamStats};


pub struct Connection {
    pub remote_addr: SocketAddr,
    pub state: ConnectionState,
    pub config: StreamConfig,
    pub stats: StreamStats,
    pub sequence_number: AtomicU32,
    pub connected_at: Option<Instant>,
    pub last_activity: Instant,
    pub ssrc: u32,
}

impl Connection {
    pub fn new(remote_addr: SocketAddr) -> Self {
        Self {
            remote_addr,
            state: ConnectionState::Disconnected,
            config: StreamConfig::default(),
            stats: StreamStats::default(),
            sequence_number: AtomicU32::new(0),
            connected_at: None,
            last_activity: Instant::now(),
            ssrc: 0,
        }
    }

    pub fn next_seq(&self) -> u32 {
        self.sequence_number.fetch_add(1, Ordering::SeqCst)
    }

    pub fn current_seq(&self) -> u32 {
        self.sequence_number.load(Ordering::SeqCst)
    }

    pub fn mark_activity(&mut self) {
        self.last_activity = Instant::now();
    }

    pub fn elapsed_since_activity(&self) -> Duration {
        self.last_activity.elapsed()
    }

    pub fn set_state(&mut self, state: ConnectionState) {
        self.state = state;
        if state == ConnectionState::Connected {
            self.connected_at = Some(Instant::now());
        }
    }

    pub fn record_sent(&mut self, bytes: usize) {
        self.stats.packets_sent += 1;
        self.stats.bytes_sent += bytes as u64;
    }

    pub fn record_received(&mut self, bytes: usize) {
        self.stats.packets_received += 1;
        self.stats.bytes_received += bytes as u64;
    }

    pub fn record_loss(&mut self, count: u64) {
        self.stats.packets_lost += count;
    }
}

pub struct ConnectionPool {
    connections: Vec<Connection>,
    max_connections: usize,
}

impl ConnectionPool {
    pub fn new(max: usize) -> Self {
        Self {
            connections: Vec::with_capacity(max),
            max_connections: max,
        }
    }

    pub fn get(&self, addr: SocketAddr) -> Option<&Connection> {
        self.connections.iter().find(|c| c.remote_addr == addr)
    }

    pub fn get_mut(&mut self, addr: SocketAddr) -> Option<&mut Connection> {
        self.connections.iter_mut().find(|c| c.remote_addr == addr)
    }

    pub fn add(&mut self, conn: Connection) -> bool {
        if self.connections.len() >= self.max_connections {
            return false;
        }
        self.connections.push(conn);
        true
    }

    pub fn remove(&mut self, addr: SocketAddr) {
        self.connections.retain(|c| c.remote_addr != addr);
    }

    pub fn len(&self) -> usize {
        self.connections.len()
    }

    pub fn iter(&self) -> impl Iterator<Item = &Connection> {
        self.connections.iter()
    }

    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut Connection> {
        self.connections.iter_mut()
    }

    pub fn contains(&self, addr: SocketAddr) -> bool {
        self.connections.iter().any(|c| c.remote_addr == addr)
    }
}

File: ports/python/rushaudio/connection.py
"""Per-peer connection tracking and a bounded connection pool.

Mirrors ``src/transport/connection.rs`` in the Rust reference implementation.
"""

import time
from dataclasses import dataclass, field
from typing import Hashable, Iterator, List, Optional

from .types import ConnectionState, StreamConfig, StreamStats


@dataclass
class Connection:
    """State and statistics for one remote peer."""

    remote_addr: Hashable
    state: ConnectionState = ConnectionState.DISCONNECTED
    config: StreamConfig = field(default_factory=StreamConfig)
    stats: StreamStats = field(default_factory=StreamStats)
    sequence_number: int = 0
    connected_at: Optional[float] = None
    last_activity: float = field(default_factory=time.monotonic)
    ssrc: int = 0

    def next_seq(self) -> int:
        """Return the current sequence number and advance it (mod 2^32)."""
        value = self.sequence_number
        self.sequence_number = (self.sequence_number + 1) & 0xFFFFFFFF
        return value

    def current_seq(self) -> int:
        return self.sequence_number

    def mark_activity(self) -> None:
        self.last_activity = time.monotonic()

    def elapsed_since_activity(self) -> float:
        return time.monotonic() - self.last_activity

    def set_state(self, state: ConnectionState) -> None:
        self.state = state
        if state == ConnectionState.CONNECTED:
            self.connected_at = time.monotonic()

    def record_sent(self, size: int) -> None:
        self.stats.packets_sent += 1
        self.stats.bytes_sent += size

    def record_received(self, size: int) -> None:
        self.stats.packets_received += 1
        self.stats.bytes_received += size

    def record_loss(self, count: int) -> None:
        self.stats.packets_lost += count


class ConnectionPool:
    """A bounded collection of connections keyed by remote address."""

    def __init__(self, max_connections: int) -> None:
        self._connections: List[Connection] = []
        self._max_connections = max_connections

    def get(self, addr: Hashable) -> Optional[Connection]:
        for conn in self._connections:
            if conn.remote_addr == addr:
                return conn
        return None

    def get_mut(self, addr: Hashable) -> Optional[Connection]:
        return self.get(addr)

    def add(self, conn: Connection) -> bool:
        if len(self._connections) >= self._max_connections:
            return False
        self._connections.append(conn)
        return True

    def remove(self, addr: Hashable) -> None:
        self._connections = [c for c in self._connections if c.remote_addr != addr]

    def len(self) -> int:
        return len(self._connections)

    def __len__(self) -> int:
        return len(self._connections)

    def is_empty(self) -> bool:
        return not self._connections

    def iter(self) -> Iterator[Connection]:
        return iter(self._connections)

    def __iter__(self) -> Iterator[Connection]:
        return iter(self._connections)

    def contains(self, addr: Hashable) -> bool:
        return any(c.remote_addr == addr for c in self._connections)